opt: Fix and improve zone-based index selection

andy-kimball opened the pull request:
This PR fixes a couple of related issues with zone-based index selection.

First, placeholder zones were incorrectly discarded. Placeholder zones are
used when there are indexes with Constraints/Leaseholder Preferences on a
table that doesn't have either of those things. The fix is to merge the
indexes placeholder into the table zone.

Second, the optimizer required that a prefix of the locality tiers match
constraints. That is no longer required after this commit. For example, if
locality=region=us;dc=east, then it now matches any of these constraint
sets equally well:

  [+locality=region=us;+dc=east]
  [+dc=east]
  [+dc=east,+dc=west]

A missing constraint match is OK (i.e. region=us), as long as a more specific
locality tier matches (i.e. dc=east).

Fixes #36642
Fixes #36644

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36689)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/config/system.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc40mEtAfgJK1YJEAfh:-Lc40mEtAfgJK1YJEAfi:b-1tdf95) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/config/system.go#L273)):*
> ```Go
> }
> 
> // getCachedZone returns the ZoneConfig for the given object ID. In the fast
> ```

Great change here, I was worried we'd have to add some special cases for our usecase, but this actually cleans up existing code!
___
*[pkg/config/system.go, line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc4-paQABDh6HDcnmGV:-Lc4-paQABDh6HDcnmGW:b-amlxap) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/config/system.go#L298)):*
> ```Go
> 			// Merge placeholder with zone by copying over subzone information.
> 			// Placeholders should only define the Subzones and SubzoneSpans fields.
> 			copy := *zone
> ```

[nit] zoneCopy (the syntax highlighting is annoying)
___
*[pkg/sql/opt/xform/coster.go, line 577 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc41iaj2PezKZlSjPFr:-Lc41iak8pKyMnann258:b-h41v5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/sql/opt/xform/coster.go#L577)):*
> ```Go
> //   0.5 = [+region=us]           // 1st tier matches
> //   0.5 = [+region=us,-dc=east]  // 1st tier matches, 2nd tier PROHIBITED
> //   0.5 = [+region=eu,+dc=east]  // 2nd tier matches, but 1st tier doesn't
> ```

Wouldn't we want to stop if we find a key that doesn't match (getting 0 here?) Seems wrong to award points for matching a `dc` in the wrong region.

`+region=eu` should be logically equivalent to `-region=us` if we have just these two regions; but they are scored differently.

I would define "match" for a given tier as:
 - 0 = key not present in constraint
 - +1 = key present in constraint and a) constraint is required and value matches; or b) constraint is prohibited and value doesn't match*
 - -1 = otherwise

[*a minor detail - if the key shows up multiple times in the constraint, (a) or (b) must hold for all instances; this would be useful in practice for something like `[-region=eu,-region=asia]` which should match `us` but not match `eu` or `asia`]

Then the score would be the length of the longest prefix that ends in `+1` and doesn't contain a `-1`.

In any case, please add the examples `[-region=eu]` and `[-region=eu,+dc=east]` above and to the test.
___
*[pkg/sql/opt/xform/coster.go, line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc45kg41jUws9hO9mw9:-Lc45kg56PMe3ZUAJjN6:b-rsw1ye) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/sql/opt/xform/coster.go#L600)):*
> ```Go
> 	}
> 
> 	// matchConstraints returns the constraint set score described in the method
> ```

not quite, it returns 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/xform/coster.go, line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc45kg41jUws9hO9mw9:-Lc7AKfXE5O2oOX47nEw:b-45e58o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/sql/opt/xform/coster.go#L600)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

not quite, it returns 
</blockquote></details>

means to say it returns what's described above as `(m-x)` (I believe)


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I left some comments on the scoring, but they have relatively minor practical implications. I'm definitely ok with this merging as-is, we really want the zone config fix.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-:-Lc7BlMbAXUjgDKkvlHq:b-blv44x)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @justinj)

___
*[pkg/sql/opt/xform/coster.go, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7O2drEQa4LT2SKP-i:-Lc7O2drEQa4LT2SKP-j:b-qxykk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/sql/opt/xform/coster.go#L606)):*
> ```Go
> 		noMatchCount := 0
> 		for i := range locality.Tiers {
> 			tier := locality.Tiers[i]
> ```

[nit] maybe you meant `tier := &locality.Tiers[i]`? If not, why not use the syntax `for i, tier := range locality.Tiers`?


<!-- Sent from Reviewable.io -->

andy-kimball committed:
opt: Fix and improve zone-based index selection

This PR fixes a couple of related issues with zone-based index selection.

First, placeholder zones were incorrectly discarded. Placeholder zones are
used when there are indexes with Constraints/Leaseholder Preferences on a
table that doesn't have either of those things. The fix is to merge the
indexes placeholder into the table zone.

Second, the optimizer required that a prefix of the locality tiers match
constraints. That is no longer required after this commit. For example, if
locality=region=us;dc=east, then it now matches any of these constraint
sets equally well:

  [+locality=region=us;+dc=east]
  [+dc=east]
  [+dc=east,+dc=west]

A missing constraint match is OK (i.e. region=us), as long as a more specific
locality tier matches (i.e. dc=east).

Fixes #36642
Fixes #36644

Release note: None

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj and @rytaft)

___
*[pkg/sql/opt/xform/coster.go, line 577 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc41iaj2PezKZlSjPFr:-Lc7OhNdEhKeHZGxGBdO:b-71m0mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/sql/opt/xform/coster.go#L577)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Wouldn't we want to stop if we find a key that doesn't match (getting 0 here?) Seems wrong to award points for matching a `dc` in the wrong region.

`+region=eu` should be logically equivalent to `-region=us` if we have just these two regions; but they are scored differently.

I would define "match" for a given tier as:
 - 0 = key not present in constraint
 - +1 = key present in constraint and a) constraint is required and value matches; or b) constraint is prohibited and value doesn't match*
 - -1 = otherwise

[*a minor detail - if the key shows up multiple times in the constraint, (a) or (b) must hold for all instances; this would be useful in practice for something like `[-region=eu,-region=asia]` which should match `us` but not match `eu` or `asia`]

Then the score would be the length of the longest prefix that ends in `+1` and doesn't contain a `-1`.

In any case, please add the examples `[-region=eu]` and `[-region=eu,+dc=east]` above and to the test.
</blockquote></details>

OK, I made changes along these lines. Also, I defined the "match" like this:
```
	//    0 = key not present in constraint set or key only matches prohibited
	//        constraints where value doesn't match
	//   +1 = key matches any required constraint key + value
	//   -1 = otherwise
```
I think (b) as you stated it is not quite right, because we only want a +1 if there was a "full" match. We want to treat prohibited constraint with non-matching value as if it was not present.
___
*[pkg/sql/opt/xform/coster.go, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7O2drEQa4LT2SKP-i:-Lc7WYcaDSwX-9ggytzs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3eac8308bd23fed3bef3e1e465a9c78d101259a/pkg/sql/opt/xform/coster.go#L606)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] maybe you meant `tier := &locality.Tiers[i]`? If not, why not use the syntax `for i, tier := range locality.Tiers`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-:-Lc7qNh04yJF4pnJ7uDq:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @justinj)

___
*[pkg/config/system.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7oYPf8DgBfbQi5jIG:-Lc7oYPf8DgBfbQi5jIH:b-wgzomg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff3d01b43bd7dd8776eec75fe0e42bf71df1ac5c/pkg/config/system.go#L57)):*
> ```Go
> 	// placeholder, there appear to be no guarantees that there can be no overlap.
> 	//
> 	// TOD(andyk): Use the combined value everywhere early in 19.2, so there's
> ```

[nit] TOD -> TODO


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-:-Lc7yaKw0xApnEGZcEAP:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/xform/coster.go, line 675 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7w6GjB6QsKmdYi1xL:-Lc7w6Gk9ckXpepoYRzF:bd5zhid) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9f72921f50596c1ee442f611a2a99ba6f1c11b/pkg/sql/opt/xform/coster.go#L675)):*
> ```Go
> 		// returned by matchConstraints for any replica. For example:
> 		//
> 		//   3: [+region=us,+dc=east]
> ```

[nit] are these 2, 1?
___
*[pkg/sql/opt/xform/coster_test.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7wRIsC0nl4rwtJPol:-Lc7wRItBAQNmGG3uleN:b-417cbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9f72921f50596c1ee442f611a2a99ba6f1c11b/pkg/sql/opt/xform/coster_test.go#L42)):*
> ```Go
> 		{locality: "region=us,dc=east", constraints: "[-region=eu]", expected: 0.0},
> 		{locality: "region=us,dc=east", constraints: "[+region=us]", expected: 0.5},
> 		{locality: "region=us,dc=east", constraints: "[+region=us,+region=eu]", expected: 0.5},
> ```

Just a note - as far as I understand `[+region=us,+region=eu]` is invalid (it's a contradiction, so no replicas would match). I don't think it matters much what this code returns in these cases though.

On the other hand multiple negative constraints do make sense, can you add a few tests like that? e.g. `[-region=asia,-region=eu]` should be 0.5 and `[-region=asia,-region=us]` should be 0.
___
*[pkg/sql/opt/xform/coster_test.go, line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7zRF486yMYUXJyOrb:-Lc7zRF486yMYUXJyOrc:bmd9yw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9f72921f50596c1ee442f611a2a99ba6f1c11b/pkg/sql/opt/xform/coster_test.go#L78)):*
> ```Go
> 
> 		var locality roachpb.Locality
> 		locality.Set(tc.locality)
> ```

linter complains about missing error checking


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/config/system.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7oYPf8DgBfbQi5jIG:-Lc89BH1BPoKw1uEod1M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff3d01b43bd7dd8776eec75fe0e42bf71df1ac5c/pkg/config/system.go#L57)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] TOD -> TODO
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/coster.go, line 675 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7w6GjB6QsKmdYi1xL:-Lc89DBx3OU_oC-XoEyx:b8c21tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9f72921f50596c1ee442f611a2a99ba6f1c11b/pkg/sql/opt/xform/coster.go#L675)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] are these 2, 1?
</blockquote></details>

Could be anything. In this case, I was thinking there are 5 replicas, where 3 are in east, 2 may not be.
___
*[pkg/sql/opt/xform/coster_test.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7wRIsC0nl4rwtJPol:-Lc85ghg0rC80AaBbsFl:b-lhq0tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9f72921f50596c1ee442f611a2a99ba6f1c11b/pkg/sql/opt/xform/coster_test.go#L42)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Just a note - as far as I understand `[+region=us,+region=eu]` is invalid (it's a contradiction, so no replicas would match). I don't think it matters much what this code returns in these cases though.

On the other hand multiple negative constraints do make sense, can you add a few tests like that? e.g. `[-region=asia,-region=eu]` should be 0.5 and `[-region=asia,-region=us]` should be 0.
</blockquote></details>

Hm, I'd assumed that when multiple keys are involved, it should function like an "OR" operator (because "AND" makes no sense), which is actually very useful functionality, since it allows you to say, "put the index in either us or eu".


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36689)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/xform/coster_test.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36689#-Lc7wRIsC0nl4rwtJPol:-Lc8AC_iCPyANeNb48El:b-26du6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9f72921f50596c1ee442f611a2a99ba6f1c11b/pkg/sql/opt/xform/coster_test.go#L42)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Hm, I'd assumed that when multiple keys are involved, it should function like an "OR" operator (because "AND" makes no sense), which is actually very useful functionality, since it allows you to say, "put the index in either us or eu".
</blockquote></details>

Since I got green build, I'm going to bors+ this. I can add those new tests in a future commit.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

andy-kimball commented:
bors r+

andy-kimball commented:
Unrelated flaky test caused bors failure.
